### PR TITLE
[clang][analyzer] Fix StreamChecker `ftell` and `fgetpos` at indeterminate file position.

### PR DESCRIPTION
**Pull request by balazske**
These functions should not be allowed if the file position is indeterminate (they return the file position).
This condition is now checked, and tests are improved to check it.

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 1]
```diff
@@ -880,6 +883,24 @@ void StreamChecker::preReadWrite(const FnDescription *Desc,
   }
 }
 
+void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
```
Comment (benshi001) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:886]: Why 　we need a separated `preWrite` ? The original `preReadWrite` also works.

Comment (balazske) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:886]: The `CallDescriptionMap` was uncomfortable to handle because too many `std::bind` functions.

Comment (steakhal) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:886]: Let me share my experience with `std::bind` and with this checker.
Unfortunately, we also have/had a couple (<10) conflicting patches to this checker, and I had to rebase 16 times to clang-18. This is fine, but when I this many `std::binds`, and sometimes changes where the railing "bool" flag meaning was changed, I really had to be on my toes when resolving syntactic & semantic conflicts.
I feel the `std::bind` usage got out of hand here, and I now regret not pushing back.

The bottom line is, that I think having dedicated handlers dispatching to common implementations, or just copy-pasting their implementation while specializing it leads to cleaner code than using `std::bind` and directly binding to the generic implementation.

Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:886]: I agree that `std::bind` is hard to follow, and dedicated handlers dispatching to common implementations is a better model.

The "Copy-pasting their implementation while specializing it" could be even better if the common parts are lifted out into helper functions; but I strongly suggest that we should avoid code duplication.

**Review by steakhal — APPROVED** [review]
I haven't gone deep into this PR, but it looks good to me.
Please wait for someone else to sign it too.

**Review by benshi001 — APPROVED** [review]
After eliminating `std::bind`, I hope there can be further solutions to reduce duplications.
